feat(toHaveCss) Overload toHaveCSS matcher to accept React.CSSProperties

iumehara opened the pull request:
References #35113

- Overload toHaveCSS to take styles property (React.CSSProperties) type.
- Extend Matchers interface in `overrides-test.d.ts` to include new signature

iumehara committed:
feat(toHaveCss) Overload toHaveCSS matcher to accept React.CSSProperties

References #35113

yury-s commented on docs/src/api/class-locatorassertions.md:
```diff
@@ -1743,6 +1751,12 @@ CSS property name.
 
 CSS property value.
 
+### param: LocatorAssertions.toHaveCSS.styles
```
Sorry, I was not clear. What I meant is you can use function overload with #1 and #2 suffixes. See `## method: TestInfo.fixme#1` for example. Then first overload will stay as is and the second one will have CSSProperty as the only parameter. Then in the documentation we'd have 2 entries, one per overload (see e.g. https://playwright.dev/docs/api/class-testinfo#test-info-fixme-1), you'd need to make sure that there is proper `usage` section that presents both signatures. Support of overloaded methods is a bit messy in our API, but CSSPropery is not much different than other overloads, so it should work.

yury-s replied:
> 1. I did not add any usage related to non-JS APIs. Is that okay? Given the new CSSProperties parameter type, I wasn't sure how to translate them to different languages.

Let's add `langs: js` to filter it out from other ports. I've added a comment.

> 2. I set since to `v1.58`. Is that right?

Yes.

iumehara committed:
Use numbered overload for CSSProperties implementation

- revert existing method definition to be the base definition. Add definition#2 that takes CSSProperties
- Define CSSProperties type that is React.CSSProperties if react is installed. If not, defaults to an object with string keys

iumehara commented on packages/playwright/src/matchers/matchers.ts:
```diff
@@ -32,6 +32,7 @@ import type { ExpectMatcherState } from '../../types/test';
 import type { TestStepInfoImpl } from '../worker/testInfo';
 import type { APIResponse, Locator, Frame, Page } from 'playwright-core';
 import type { FrameExpectParams } from 'playwright-core/lib/client/types';
+import type { CSSProperties } from 'react';
```
I added a [type](https://github.com/microsoft/playwright/pull/38617/files#diff-a0ff89a66553228f94756761698ea31af6d7498492e15cc4cb6bc3f906436d69R1752) that uses 'react' only if it is imported correctly. If not, it uses a fallback type that has string keys.

Does this look okay? Also, is this the correct place to define this. I originally defined/exported this from matchers.ts, but had trouble getting the import paths to resolve.

yury-s commented on docs/src/api/class-locatorassertions.md:
```diff
@@ -1749,6 +1749,30 @@ CSS property value.
 * since: v1.18
 
+## async method: LocatorAssertions.toHaveCSS#2
+* since: v1.20
```
let's add `* langs: js` to not have in language ports.

iumehara committed:
Update toHaveCSS to have correct overload syntax

- give the base implementation a #1 overload and update referencing documentation
- CSSProperties type defaults to never if react is not installed

iumehara committed:
Update docs/src/release-notes-js.md

Co-authored-by: Yury Semikhatsky <yurys@chromium.org>
Signed-off-by: iumehara <iumehara@users.noreply.github.com>

iumehara committed:
Update types/test.d.ts to reflect note on CSSProperties
